clarity/contracts/arkadiko-swap-v1-1.clar

```diff
@@ -0,0 +1,435 @@
+      (balance-x (get balance-x pair))
+      (balance-y (get balance-y pair))
+      (shares (unwrap-panic (contract-call? swap-token-trait get-balance tx-sender)))
+      (shares-total (get shares-total pair))
+      (withdrawal-x (/ (* shares balance-x) shares-total))
+      (withdrawal-y (/ (* shares balance-y) shares-total))
+    )
+    (ok (list withdrawal-x withdrawal-y))
+  )
+)
+
+;; ;; reduce the amount of liquidity the sender provides to the pool
+;; ;; to close, use u100
+(define-public (reduce-position (token-x-trait <ft-trait>) (token-y-trait <ft-trait>) (swap-token-trait <swap-token>) (percent uint))
+  (let
+    (
+      (token-x (contract-of token-x-trait))
+      (token-y (contract-of token-y-trait))
+      (pair (unwrap-panic (map-get? pairs-data-map { token-x: token-x, token-y: token-y })))
+      (balance-x (get balance-x pair))
+      (balance-y (get balance-y pair))
+      (shares (unwrap-panic (contract-call? swap-token-trait get-balance tx-sender)))
+      (shares-total (get shares-total pair))
+      (contract-address (as-contract tx-sender))
+      (withdrawal (/ (* shares percent) u100))
+      (withdrawal-x (/ (* withdrawal balance-x) shares-total))
+      (withdrawal-y (/ (* withdrawal balance-y) shares-total))
+      (pair-updated
+        (merge pair
+          {
+            shares-total: (- shares-total withdrawal),
+            balance-x: (- (get balance-x pair) withdrawal-x),
+            balance-y: (- (get balance-y pair) withdrawal-y)
+          }
+        )
+      )
+    )
+
+    (asserts! (<= percent u100) (err u5))
+    (asserts! (is-ok (contract-call? token-x-trait transfer withdrawal-x contract-address tx-sender none)) transfer-x-failed-err)
+    (asserts! (is-ok (contract-call? token-y-trait transfer withdrawal-y contract-address tx-sender none)) transfer-y-failed-err)
+
+    ;; (unwrap-panic (decrease-shares token-x token-y tx-sender withdrawal)) ;; should never fail, you know...
+    (map-set pairs-data-map { token-x: token-x, token-y: token-y } pair-updated)
+    (try! (contract-call? swap-token-trait burn tx-sender withdrawal))
+
+    (print { object: "pair", action: "liquidity-removed", data: pair-updated })
+    (ok (list withdrawal-x withdrawal-y))
+  )
+)
+
+;; exchange known dx of x-token for whatever dy of y-token based on current liquidity, returns (dx dy)
+;; the swap will not happen if can't get at least min-dy back
+(define-public (swap-x-for-y (token-x-trait <ft-trait>) (token-y-trait <ft-trait>) (dx uint) (min-dy uint))
+  (let (
+    (token-x (contract-of token-x-trait))
+    (token-y (contract-of token-y-trait))
+    (pair (unwrap-panic (map-get? pairs-data-map { token-x: token-x, token-y: token-y })))
+    (balance-x (get balance-x pair))
+    (balance-y (get balance-y pair))
+    (sender tx-sender)
+    (dx-with-fees (/ (* u997 dx) u1000)) ;; 0.3% fee for LPs
+    (dy (/ (* balance-y dx-with-fees) (+ balance-x dx-with-fees)))
+    (fee (/ (* u5 dx) u10000)) ;; 0.05% fee for protocol
```

Comment:
@nieldeckx Q: are you taking 0.35% now? What uniswap does is 0.25 + 0.05, I think that makes more sense?

Reply:
No, the user pays 0.3% in fees out of which the protocol can take 0.05%.

Reply:
Ok 👍